components/fxa-client/ios/FxAClient/FirefoxAccount.swift

```diff
@@ -201,109 +146,243 @@ open class FirefoxAccount {
+    }
+
+    open func setDeviceDisplayNameSync(_ name: String) throws {
+        try rustCall { err in
+            fxa_set_device_name(self.raw, name, err)
+        }
+    }
+
+    open func pollDeviceCommandsSync() throws -> [DeviceEvent] {
+        let ptr = try rustCall { err in
+            fxa_poll_device_commands(self.raw, err)
+        }
+        defer { fxa_bytebuffer_free(ptr) }
+        tryPersistState()
+        let msg = try! MsgTypes_AccountEvents(serializedData: Data(rustBuffer: ptr))
+        return DeviceEvent.fromCollectionMsg(msg: msg)
+    }
+
+    open func handlePushMessageSync(payload: String) throws -> [DeviceEvent] {
+        let ptr = try rustCall { err in
+            fxa_handle_push_message(self.raw, payload, err)
+        }
+        defer { fxa_bytebuffer_free(ptr) }
+        tryPersistState()
+        let msg = try! MsgTypes_AccountEvents(serializedData: Data(rustBuffer: ptr))
+        return DeviceEvent.fromCollectionMsg(msg: msg)
+    }
+
+    open func sendSingleTabSync(targetId: String, title: String, url: String) throws {
+        try rustCall { err in
+            fxa_send_tab(self.raw, targetId, title, url, err)
+        }
+    }
+
+    open func setDevicePushSubscriptionSync(endpoint: String, publicKey: String, authKey: String) throws {
+        try rustCall { err in
+            fxa_set_push_subscription(self.raw, endpoint, publicKey, authKey, err)
+        }
+    }
+
+    open func initializeDeviceSync(
+        name: String,
+        deviceType: DeviceType,
+        supportedCapabilities: [DeviceCapability]
+    ) throws {
+        let (data, size) = capabilitiesToBuffer(capabilities: supportedCapabilities)
+        try data.withUnsafeBytes { bytes in
+            try rustCall { err in
+                fxa_initialize_device(
+                    self.raw,
+                    name,
+                    Int32(deviceType.toMsg().rawValue),
+                    bytes.bindMemory(to: UInt8.self).baseAddress!,
+                    size,
+                    err
+                )
+            }
+        }
+    }
+
+    open func ensureCapabilitiesSync(supportedCapabilities: [DeviceCapability]) throws {
+        let (data, size) = capabilitiesToBuffer(capabilities: supportedCapabilities)
+        try data.withUnsafeBytes { bytes in
+            try rustCall { err in
+                fxa_ensure_capabilities(
+                    self.raw,
+                    bytes.bindMemory(to: UInt8.self).baseAddress!,
+                    size,
+                    err
+                )
+            }
+        }
+    }
+
+    internal func capabilitiesToBuffer(capabilities: [DeviceCapability]) -> (Data, Int32) {
+        let msg = MsgTypes_Capabilities.with {
+            $0.capability = capabilities.map { $0.toMsg() }
+        }
+        let data = try! msg.serializedData()
+        let size = Int32(data.count)
+        return (data, size)
+    }
+}
+
+// This queue serves as a semaphore to the rust layer.
```

Comment:
nit: why is that important/necessary? (I mean, I have a fair idea, but it's not immediately obvious given that the rust layer has its own locking inside the `ConcurrentHandleMap`).